# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
#Создадим таблицу, с количеством долгов, суммарным долгом и процентажом невозврата кредита
#рассчитаем долю не возврата кредита в зависимости от количества детей с помощью 'mean'
child_debt = data.pivot_table(index=['children'], values='debt', aggfunc=['sum','count','mean'])
#Переименуем стобец 'mean' в '%' для наглядности
child_debt = child_debt.rename(columns= {'mean':'%','sum':'общая сумма','count':'общее количество'})
#Умножим значения на сто для получения процента, а не доли
child_debt['%'] = (child_debt['%']*100).round(2)
#Выведем на экран таблицу
child_debt

общая сумма общее количество     %
                debt             debt  debt
children                                   
0               1063            14091  7.54
1                444             4808  9.23
2                194             2052  9.45
3                 27              330  8.18
4                  4               41  9.76
5                  0                9  0.00

**Анализ зависимости между количеством детей и возвратом кредита в срок.**
**Гипотеза:**
Количество детей влияет на возвращение кредита срок.
**Задачи:**
1. Определить при каком количестве детей наибольший процент случаев не возвращение кредита в срок
2. Проанализировать значения. Выявить группу риска.

**Результаты анализа**
  По преведенной таблице видно можно сделать следующие предположения:
1. Группа заемщик с количеством детей равным 5, почти всегда, закрывают задолженность в срок. Можно сделать предположение, что это значение аномально, стоит провести дальнейшее исследования и проверить оригинальные даннные. Также вероятно, что клиенты с таким колчеством клиентов почти не берут крупные кредиты или не берут вовсе, что также отражает приведенная таблица выше.
2. Наибольший процент непогашенных кредитов у заемщик с 4 детьми, в этом же диапазоне значений и находятся заемщики с 1 или 2 детьми. Можно объединить их в группу: 'хай риск'
3. Процент непогашенных кредитов у заемщик с 0;3 детьми находятся в близком дипазоне, их можно объединить в группу: 'мид риск'

  Действительно, существует зависимость между количеством детей и возвращением кредита. Стоит заметить, что больше всего берут крдеты клиенты в группе с количеством детей от 2-0, В данной группе наибольший процент непогашенных кредитов у клиентов с 1 или 2 детьми

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
#Создадим таблицу, с количеством долгов, суммарным долгом и процентажом невозврата кредита
#рассчитаем долю не возврата кредита в зависимости от количества детей с помощью 'mean'
family_status_debt = data.pivot_table(index=(['family_status']), values='debt', aggfunc=['sum','count','mean'])
#Переименуем стобец 'mean' в '%' для наглядности
family_status_debt = family_status_debt.rename(columns= {'mean':'%','sum':'общая сумма','count':'общее количество'})
#Умножим значения на сто для получения процента, а не доли
family_status_debt['%'] = (family_status_debt['%']*100).round(2)
#Выведем на экран таблицу
family_status_debt

общая сумма общее количество     %
                             debt             debt  debt
family_status                                           
Не женат / не замужем         273             2796  9.76
в разводе                      84             1189  7.06
вдовец / вдова                 63              951  6.62
гражданский брак              385             4134  9.31
женат / замужем               927            12261  7.56

**Анализ зависимости между семейныйм положением и возвратом кредита в срок.**
**Гипотеза:**
Семейное положение влияет на возвращение кредита срок.
**Задачи:**
1. Определить при каком семейном положении наибольший процент случаев не возвращение кредита в срок
2. Проанализировать значения. Выявить закономерности.

**Результаты анализа**
1. Люди, находящиеся в гражданском браке или не состоящие в зарегистрированном браке, имеют более высокий процент невозврата кредита в срок по сравнению с остальными категориями. Это может указывать на то, что отсутствие официального брачного статуса может повлиять на финансовую стабильность и ответственность перед кредиторами.

2. Люди, находящиеся в разводе или являющиеся вдовцами / вдовами, имеют более низкий процент невозврата кредита в срок по сравнению с остальными категориями. Это может объясняться повышенной финансовой осведомленностью и ответственностью, возникающей при разводе или потере супруга.

3. Женатые / замужние люди имеют средний процент невозврата кредита в срок

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
#Создадим таблицу, с количеством долгов, суммарным долгом и процентажом невозврата кредита('mean')
#и распределим по категориям дохода, полученным выше.
income_debt = data.pivot_table(index = ['total_income_category'], values='debt',aggfunc=['sum','count','mean'])
#Переименуем стобец 'mean' в '%' для наглядности
income_debt = income_debt.rename(columns={'mean':'%','sum':'общая сумма','count':'общее количество'})
#Умножим значения на сто для получения процента, а не доли
income_debt['%'] = (income_debt['%']*100).round(2)
#Выведем на экран таблицу
income_debt

общая сумма общее количество     %
                             debt             debt  debt
total_income_category                                   
A                               2               25  8.00
B                             354             5014  7.06
C                            1353            15921  8.50
D                              21              349  6.02
E                               2               22  9.09

**Анализ зависимости между уровнем дохода и возвратом кредита в срок.**
**Гипотеза:**
Уровень дохода влияет на возвращение кредита срок.
**Задачи:**
1. Определить в какой категории дохода клиентов наибольший процент невозврат кредита в срок.
2. Проанализировать значения. Выявить закономерности.

**Результаты анализа**
1. У клиентов с доходностью в категории С наибольшее число невозращенных кредитов,процент невозврата относитльно всего количества достаточно высокий.
2. Примерно в одном диапазоне по количеству, находится число не возвратов кредитов, которые были взяты клиентами с доходностью в категории А и Е, стоит заметить, что в этом диапазоне и находятся те цели, при которых процент невозвращения кредита относительно количества различный, у категории А - 8%, у Е-9%, который является наибольшим в объекте исследования.
3. Количество невозвращенных кредитов, взятых клинтами с доходностью В и D маленькое, и процент невозвращения тоже маленький.
Можно сделать вывод, что наибольший процент невозращения кредита в срок у клиентов с доходностью в категории Е, С, А(категории расположены в порядке убывания процента). Стоит заметить, что количество невозвращенных кредитов в категории С много больше, чем в А и Е, а сам процент, достаточно на высоком уровне, 2 по величине, стоит именну эту категорию как ключевой при исследовании зависимости не возвращения кредита от дохода. Прямой зависимости, выявить неудалось из-за различий значений, цдалось выявить ключевую группу.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
#Создадим таблицу, с количеством долгов, суммарным долгом и процентажом невозврата кредита('mean')
#и распределим по категориям целей займы(purpose_category), полученным выше
purpose_debt = data.pivot_table(index = ['purpose_category'], values='debt',aggfunc=['sum','count','mean'])
#Переименуем стобец 'mean' в '%' для наглядности
purpose_debt = purpose_debt.rename(columns={'mean':'%','sum':'общая сумма','count':'общее количество'})
#Умножим значения на сто для получения процента, а не доли
purpose_debt['%'] = (purpose_debt['%']*100).round(2)
#Выведем на экран таблицу
purpose_debt

общая сумма общее количество     %
                                debt             debt  debt
purpose_category                                           
операции с автомобилем           400             4279  9.35
операции с недвижимостью         780            10751  7.26
получение образования            369             3988  9.25
проведение свадьбы               183             2313  7.91

**Анализ зависимости между целью кредита и возвратом кредита в срок.**
**Гипотеза:**
Цель кредита влияет на возвращение кредита срок.
**Задачи:**
1. Определить при какой цели получения кредита у клиентов наибольший процент невозврат кредита в срок.
2. Проанализировать значения. Выявить закономерности.

**Результаты анализа**
1. При опреациях с недвижимостью наибольшее число невозращенных кредитов, но наименьший процент невозврата относитльно всего количества.
2. Примерно в одном диапазоне по количеству, находится число не возвратов кредитов, которы были взяты из-за:получение образования и операции с автомобилем, стоит заметить, что в этом диапазоне и находятся те цели, при которых процент невозвращения кредита относительно количества наибольший.
3. Количество невозвращенных кредитов, взятых ввиду провдения свадьбы наименьшее, а также процент невозвращения крдеит относительно количества тоже маленький, при рассмотрении остальных категорий
Удалось зависимость, которая гласит, что процент не возвращения кредита наибольший у клиентов со следущими целями: операции с автомобилем, получение образования; процент не возвращения кредита наименьший у клиентов со следущими целями: операции с недвижимостью, проведение свадьбы.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
1. Человеческий фактор: Пропуски могут возникать из-за ошибок ввода данных, незаполненных полей или неправильной обработки информации операторами ввода данных. Это может произойти, когда информация собирается вручную и требуется ввод данных вручную.

2. Технические неполадки: Иногда пропуски могут возникать из-за технических проблем, таких как сбои в системе, потеря данных или неполадки при передаче информации. Это может быть связано с неполадками в сети, проблемами с серверами или программными сбоями.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Количественные перменные можно сравнить друг с другом. В них могут  быть аномально большие занчения, также аномально маленькие, среднее арфметическое будет сильно искажено этими экстремальными значениями. А медиана покажет "среднее по палате".

In [26]:
# Пример к комментарию ревьюера
from numpy import median, mean
a = [1, 1, 1, 1, 1, 1, 1, 1, 100]
print('Медиана:', median(a))
print('Среднее:', mean(a))

Медиана: 1.0
Среднее: 12.0


### Шаг 4: общий вывод.

In [27]:
frames = [purpose_debt, income_debt, family_status_debt, child_debt]
#сделаем общую таблицу
result = pd.concat(frames)
result

общая сумма общее количество     %
                                debt             debt  debt
операции с автомобилем           400             4279  9.35
операции с недвижимостью         780            10751  7.26
получение образования            369             3988  9.25
проведение свадьбы               183             2313  7.91
A                                  2               25  8.00
B                                354             5014  7.06
C                               1353            15921  8.50
D                                 21              349  6.02
E                                  2               22  9.09
Не женат / не замужем            273             2796  9.76
в разводе                         84             1189  7.06
вдовец / вдова                    63              951  6.62
гражданский брак                 385             4134  9.31
женат / замужем                  927            12261  7.56
0                               1063            14091  7.54
1                                444             4808  9.23
2                                194             2052  9.45
3                                 27              330  8.18
4                                  4               41  9.76
5                                  0                9  0.00

**Выводы**  
  * При проведении исследования предоставленой банком статистики о платёжеспособности клиентов были выявлены зависимости между:
семейным положением клиента, количеством детей на факт погашения кредита в срок и целью кредита. 
  * Максимальный процент не возврата кредита у клиентов: с семейным положением "Не женат / не замужем" и "гражданский брак" (9,76 и 9,31 соответственно), а также у клиентов с 4 детьми (9.76), 1 ребенком(9.23),с 2 детьми (9.45),и с такими целями как операции с автомобилем, получение образования. 
  * Клиенты, входящие в данные категории являются клиентами с высокой степенью риска не возврата кредита. Пересечение данных категорий с высокой вероятностью приведет к не возвращение кредита. 